x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/perform_rule_upgrade/create_upgradeable_rules_payload.ts

```diff
@@ -0,0 +1,190 @@
+import { determineFieldUpgradeStatus } from './determine_field_upgrade_status';
+import { FIELDS_PAYLOAD_BY_RULE_TYPE } from './create_fields_payload_by_rule_type';
+import { RULE_DEFAULTS } from '../../../rule_management/logic/detection_rules_client/mergers/apply_rule_defaults';
+import { calculateRuleFieldsDiff } from '../../logic/diff/calculation/calculate_rule_fields_diff';
+import { convertPrebuiltRuleAssetToRuleResponse } from '../../../rule_management/logic/detection_rules_client/converters/convert_prebuilt_rule_asset_to_rule_response';
+
+const RULE_DEFAULTS_FIELDS_SET = new Set(Object.keys(RULE_DEFAULTS));
+
+interface BuildUpgradePayloadProps {
+  upgradeableRules: UpgradeableRule[];
+  requestBody: PerformRuleUpgradeRequestBody;
+}
+
+export const createUpgradeableRulesPayload = ({
+  upgradeableRules,
+  requestBody,
+}: BuildUpgradePayloadProps) => {
+  const { pick_version: globalPickVersion = PickVersionValuesEnum.MERGED } = requestBody;
+  const ruleUpgradePayloads: PrebuiltRuleAsset[] = [];
+  const payloadErrors: Array<PromisePoolError<{ rule_id: string }>> = [];
+
+  upgradeableRules.forEach((upgradeableRule) => {
+    const targetRuleType = upgradeableRule.target.type;
+    const ruleId = upgradeableRule.target.rule_id;
+    const fieldsPayload = FIELDS_PAYLOAD_BY_RULE_TYPE.get(targetRuleType);
+
+    try {
+      if (fieldsPayload === undefined) {
+        throw new Error(`Unexpected rule type: ${targetRuleType}`);
+      }
+
+      assertRuleTypeMatch({ upgradeableRule, requestBody });
+
+      const calculatedRuleDiff = calculateRuleFieldsDiff({
+        base_version: upgradeableRule.base
+          ? convertRuleToDiffable(convertPrebuiltRuleAssetToRuleResponse(upgradeableRule.base))
+          : MissingVersion,
+        current_version: convertRuleToDiffable(upgradeableRule.current),
+        target_version: convertRuleToDiffable(
+          convertPrebuiltRuleAssetToRuleResponse(upgradeableRule.target)
+        ),
+      }) as AllFieldsDiff;
+
+      let rulePayload: Partial<PrebuiltRuleAsset> = {};
+
+      for (const field of fieldsPayload) {
+        rulePayload = {
+          ...rulePayload,
+          [field]: getValueForField({
+            field,
+            upgradeableRule,
+            globalPickVersion,
+            requestBody,
+            ruleId,
+            ruleFieldsDiff: calculatedRuleDiff,
+          }),
+        };
+      }
+
+      ruleUpgradePayloads.push(rulePayload as PrebuiltRuleAsset);
+    } catch (err) {
+      payloadErrors.push({
+        error: err,
+        item: { rule_id: ruleId },
+      });
+    }
+  });
+
+  return {
+    payloadErrors,
+    ruleUpgradePayloads,
+  };
+};
+
+interface GetValueForFieldArgs {
+  field: keyof PrebuiltRuleAsset;
+  upgradeableRule: UpgradeableRule;
+  globalPickVersion: PickVersionValues;
+  requestBody: PerformRuleUpgradeRequestBody;
+  ruleId: RuleSignatureId;
+  ruleFieldsDiff: AllFieldsDiff;
+}
+
+const getValueForField = ({
```

Comment:
I think this function should be refactored and simplified. It accepts too many params while `getGroupNameForField(field, upgradeableRule.target.type);` invoked twice under the hood and its result should be used as a param.

I'd say this function is only concerned by a field so it should accept `fieldName` and field triad. `ruleId` shouldn't be a part of params. Error messages could be enriched by using folded `try`/`catch` blocks.

Reply:
I agree we can remove `ruleId`, just did. 

Not sure I understand your proposal for other params. In order to remove them, we'd have to inline most of the logic  into `createUpgradeableRulesPayload`. I think it's more clear this way: all the logic for retrieving the value for a single field is within this function. Once that function ends running, we go the next field to retrieve its value. And so on.